File: src/test/java/org/artembogomolova/demo/webapp/test/domain/entity/core/PersonEntityTest.java
package org.artembogomolova.demo.webapp.test.domain.entity.core;

import java.util.Calendar;
import java.util.Date;
import java.util.List;
import lombok.EqualsAndHashCode;
import org.artembogomolova.demo.webapp.domain.IdentifiedEntity;
import org.artembogomolova.demo.webapp.domain.core.Person;
import org.artembogomolova.demo.webapp.domain.core.Person_;
import org.artembogomolova.demo.webapp.domain.core.PhysicalAddress;
import org.artembogomolova.demo.webapp.test.domain.DomainTestUtil;
import org.artembogomolova.demo.webapp.test.domain.entity.AbstractAccessorEntityTest;
import org.junit.jupiter.api.Assertions;
import org.junit.jupiter.api.DisplayName;

@DisplayName("Entity test: Person")
class PersonEntityTest extends AbstractAccessorEntityTest<Person> {

  private static final String NAME_VALUE = "TestName";
  private static final String SURNAME_VALUE = "TestSurname";
  private static final String PATRONYMIC_VALUE = "TestPatronymic";
  private static final Date BIRTH_DATE_VALUE = Calendar.getInstance().getTime();
  private static final String PHONE_VALUE = "+7-812-000-00-00";
  private static final String EMAIL_VALUE = "test_email@gmail.com";
  private static final PhysicalAddress PHYSICAL_ADDRESS_VALUE = DomainTestUtil.buildTestAddress();

  PersonEntityTest() {
    super(Person.class,
        Person::new,
        MockPerson::new);
  }

  @Override
  protected Person buildStandardEntity() {
    Person result = new Person();
    result.setName(NAME_VALUE);
    result.setSurname(SURNAME_VALUE);
    result.setPatronymic(PATRONYMIC_VALUE);
    result.setBirthDate(BIRTH_DATE_VALUE);
    result.setPhone(PHONE_VALUE);
    result.setEmail(EMAIL_VALUE);
    result.setEstateAddress(PHYSICAL_ADDRESS_VALUE);
    return result;
  }

  @Override
  protected void containFieldCorrectValuesTest(Person standardEntity) {
    Assertions.assertEquals(NAME_VALUE, standardEntity.getName());
    Assertions.assertEquals(SURNAME_VALUE, standardEntity.getSurname());
    Assertions.assertEquals(PATRONYMIC_VALUE, standardEntity.getPatronymic());
    Assertions.assertEquals(BIRTH_DATE_VALUE, standardEntity.getBirthDate());
    Assertions.assertEquals(PHONE_VALUE, standardEntity.getPhone());
    Assertions.assertEquals(EMAIL_VALUE, standardEntity.getEmail());
    Assertions.assertEquals(PHYSICAL_ADDRESS_VALUE, standardEntity.getEstateAddress());
    Assertions.assertEquals(null, standardEntity.getUser());
    Assertions.assertTrue(standardEntity.getOrders().isEmpty());
  }

  @Override
  protected List<String> getAvailableConstraintNames() {
    return List.of(IdentifiedEntity.BASIC_CONSTRAINT_NAME, Person.PHONE_CONSTRAINT_NAME,
        Person.EMAIL_CONSTRAINT_NAME);
  }


  @Override
  protected Person buildAnotherEntityForTest() {
    return DomainTestUtil.buildPerson();
  }

  @Override
  protected boolean withoutBasicConstraint(Person standardEntity, String columnName) {
    if (Person_.BIRTH_DATE.equals(columnName)) {
      withoutColumnEqualTest(standardEntity, Person::getBirthDate, Person::setBirthDate);
      return true;
    }
    return personNamesBlockTest(standardEntity, columnName);
  }

  private boolean personNamesBlockTest(Person standardEntity, String columnName) {
    switch (columnName) {
      case Person_.NAME: {
        withoutColumnEqualTest(standardEntity, Person::getName, Person::setName);
        return true;
      }
      case Person_.PATRONYMIC: {
        withoutColumnEqualTest(standardEntity, Person::getPatronymic, Person::setPatronymic);
        return true;
      }
      case Person_.SURNAME: {
        withoutColumnEqualTest(standardEntity, Person::getSurname, Person::setSurname);
        return true;
      }
      default:
        return false;
    }
  }

  protected boolean withoutAlternateConstraints(Person standardEntity, String constraintName, String columnName) {
    switch (constraintName) {
      case Person.PHONE_CONSTRAINT_NAME: {
        withoutColumnEqualTest(standardEntity, Person::getPhone, Person::setPhone);
        return true;
      }
      case Person.EMAIL_CONSTRAINT_NAME: {
        withoutColumnEqualTest(standardEntity, Person::getEmail, Person::setEmail);
        return true;
      }
      default:
        return false;
    }
  }

  @EqualsAndHashCode(callSuper = false)
  private static class MockPerson extends Person {

    MockPerson(Person person) {
      super(person);
    }
  }
}

File: app/src/test/kotlin/org/artembogomolova/demo/webapp/test/domain/core/PersonEntityTest.kt
package org.artembogomolova.demo.webapp.test.domain.core

import java.util.Calendar
import java.util.Date
import org.artembogomolova.demo.webapp.main.domain.core.Person
import org.artembogomolova.demo.webapp.main.domain.core.Person_
import org.artembogomolova.demo.webapp.main.domain.core.PhysicalAddress
import org.artembogomolova.demo.webapp.test.domain.AbstractAccessorEntityTest
import org.junit.jupiter.api.Assertions
import org.junit.jupiter.api.DisplayName

@DisplayName("Entity test: Person")
internal class PersonEntityTest :
    AbstractAccessorEntityTest<Person>(
        Person::class.java,
        Person::from
    ) {
    override fun buildStandardEntity(): Person = Person(
        name = NAME_VALUE,
        surname = SURNAME_VALUE,
        patronymic = PATRONYMIC_VALUE,
        phone = PHONE_VALUE,
        estateAddress = PHYSICAL_ADDRESS_VALUE,
        birthDate = BIRTH_DATE_VALUE.time
    )


    override fun containFieldCorrectValuesTest(standardEntity: Person) {
        Assertions.assertEquals(NAME_VALUE, standardEntity.name)
        Assertions.assertEquals(SURNAME_VALUE, standardEntity.surname)
        Assertions.assertEquals(PATRONYMIC_VALUE, standardEntity.patronymic)
        Assertions.assertEquals(BIRTH_DATE_VALUE.time, standardEntity.birthDate)
        Assertions.assertEquals(PHONE_VALUE, standardEntity.phone)
        Assertions.assertEquals(PHYSICAL_ADDRESS_VALUE, standardEntity.estateAddress)
        Assertions.assertEquals(null, standardEntity.user)
    }

   override fun buildAnotherEntityForTest(): Person {
        val result = buildStandardEntity()
        result.name = "ABC"
        return result
    }

    override fun withoutBasicConstraint(standardEntity: Person, columnName: String): Boolean {
        if (Person_.BIRTH_DATE == columnName) {
            withoutColumnEqualTest(standardEntity, Person::birthDate)
            return true
        }
        return personNamesBlockTest(standardEntity, columnName)
    }

    private fun personNamesBlockTest(standardEntity: Person, columnName: String): Boolean {
        return when (columnName) {
            Person_.NAME -> {
                withoutColumnEqualTest(standardEntity, Person::name)
                true
            }
            Person_.PATRONYMIC -> {
                withoutColumnEqualTest(standardEntity, Person::patronymic)
                true
            }
            Person_.SURNAME -> {
                withoutColumnEqualTest(standardEntity, Person::surname)
                true
            }
            else -> false
        }
    }

    companion object {
        private const val NAME_VALUE = "TestName"
        private const val SURNAME_VALUE = "TestSurname"
        private const val PATRONYMIC_VALUE = "TestPatronymic"
        private val BIRTH_DATE_VALUE: Date = Calendar.getInstance().time
        private const val PHONE_VALUE = "+7-812-000-00-00"
        private val PHYSICAL_ADDRESS_VALUE: PhysicalAddress? = PhysicalAddress(null, null, null, null)
    }
}